adc/ping.go

```diff
@@ -0,0 +1,199 @@
+package adc
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/direct-connect/go-dcpp/adc/types"
+)
+
+type PingHubInfo struct {
+	Ext     []string
+	HubInfo HubInfo
+	Users   []User
+}
+
+//ctx context.Context
+
+func Ping(ctx context.Context, addr string) (*PingHubInfo, error) {
+	if i := strings.Index(addr, "://"); i >= 0 {
+		proto := addr[:i]
+		addr = addr[i+3:]
+		switch proto {
+		case SchemaADC, SchemaADCS:
+			// continue
+		default:
+			return nil, fmt.Errorf("unsupported protocol: %q", proto)
+		}
+	}
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	c, err := NewConn(conn)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+
+	fea := ModFeatures{
+		// should always be set for ADC
+		FeaBASE: true,
+		FeaBAS0: true,
+		FeaTIGR: true,
+		// extensions
+
+		// TODO: some hubs will stop the handshake after sending the hub info
+		//       if this extension is specified
+		//adc.FeaPING: true,
+
+		FeaBZIP: true,
+		// TODO: ZLIG
+	}
+
+	err = c.WriteHubMsg(Supported{
+		Features: fea,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Flush(); err != nil {
+		return nil, err
+	}
+
+	// set deadline once
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(time.Second * 10)
+	}
+	if err = c.conn.SetDeadline(deadline); err != nil {
+		return nil, err
+	}
+
+	var hub PingHubInfo
+
+	// first, we expect a SUP from the hub with a list of supported features
+	msg, err := c.ReadInfoMsg(deadline)
+	if err != nil {
+		return nil, err
+	}
+	sup, ok := msg.(Supported)
+	if !ok {
+		return nil, fmt.Errorf("expected SUP command, got: %#v", msg)
+	}
+	var ext []Feature
+	var f []string
+	for features, ok := range sup.Features {
+		if !ok {
+			continue
+		}
+		ext = append(ext, features)
+		f = append(f, features.String())
+	}
+	hub.Ext = f
+
+	// next, we expect a SID that will assign a Session ID
+	msg, err = c.ReadInfoMsg(deadline)
+	if err != nil {
+		return nil, err
+	}
+	sid, ok := msg.(SIDAssign)
+	if !ok {
+		return nil, fmt.Errorf("expected SID command, got: %#v", msg)
+	}
+
+	var user User
+	pid := types.NewPID()
+	user.Pid = &pid
+	user.Id = user.Pid.Hash()
+	num := int64(time.Now().Nanosecond())
+	user.Name = "pinger_" + strconv.FormatInt(num, 16)
+	user.Features = ext
+	user.Slots = 1
+
+	err = c.WriteBroadcast(sid.SID, user)
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Flush(); err != nil {
+		return nil, err
+	}
+
+	if err = c.conn.SetDeadline(deadline); err != nil {
```

Comment:
It looks like this doesn't make sense here - it will be set by `ReadPacket(deadline)` below.